### PR TITLE
Add socket connection timeout.

### PR DESCRIPTION
**Pull request by TiagoMSSantos**
The pdb doesn't have a connection timeout, so when the connection to the database fails abruptly, the threads block waiting for a response of a submitted query.
This new branch adds a connection and a login connection timeout (defined by the user) so it doesn't block the system.

### CODE REVIEW

**Comment by TiagoMSSantos** [comment]
I just want to add that this fix doesn't work with Oracle databases, so don't focus too much on the code, but rather on the approach of adding new settings to configure the socket connection timeout.

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java [hunk 1]
```diff
@@ -150,6 +152,14 @@
      * space in the DB server. This depends on the database implementation.
      */
     public static final String DISABLE_LOB_CACHING = "pdb.disable_lob_caching";
+    /**
+     * Property that indicates the time of login timeout in the socket.
+     */
+    public static final String LOGIN_TIMEOUT = "pdb.login_timeout";
+    /**
+     * Property that indicates the time of login timeout in the socket.
```
Comment (nuno-azevedo) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java]: Copy pastey from above.

File: src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java [hunk 2]
```diff
@@ -84,4 +84,12 @@
      * server. This is only possible on implementations that support this behavior. The default value is {@code false}.
      */
     public static final boolean DEFAULT_DISABLE_LOB_CACHING = false;
+    /**
+     * The default login socket timeout.
```
Comment (nuno-azevedo) [inline src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java]: login socket?

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java (hunk 1 above)
Comment (TiagoMSSantos) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java]: You are right. My bad. Thanks for mention it.

File: src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java (hunk 2 above)
Comment (TiagoMSSantos) [reply src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java]: It's the timeout of making a connection to the socket. Because, the database can fail to connect. I will describe it better in javadocs.

Comment (lmtjalves) [reply src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java]: Suggestion: Default duration (in seconds) to wait for the database connection to be established.

Comment (TiagoMSSantos) [reply src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java]: Thanks for the suggestion. You are right. That way makes more sense.

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java [hunk 3]
```diff
@@ -150,6 +152,14 @@
      * space in the DB server. This depends on the database implementation.
      */
     public static final String DISABLE_LOB_CACHING = "pdb.disable_lob_caching";
+    /**
```
Comment (lmtjalves) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java]: new line before the javadoc

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java [hunk 4]
```diff
@@ -150,6 +152,14 @@
...
+     * Property that indicates the time, in seconds, of login timeout in the socket.
+     */
+    public static final String LOGIN_TIMEOUT = "pdb.login_timeout";
+    /**
```
Comment (lmtjalves) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java]: new line before the javadoc

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java [hunk 5]
```diff
@@ -150,6 +152,14 @@
      * space in the DB server. This depends on the database implementation.
      */
     public static final String DISABLE_LOB_CACHING = "pdb.disable_lob_caching";
+    /**
+     * Property that indicates the time, in seconds, of login timeout in the socket.
```
Comment (lmtjalves) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java]: I'm OK with keeping the property name as is, but I think there's missing some more context on the javadoc. For instance, something like: "... in seconds, to wait for the database connection to be established" is more easy to understand than "login timeout in the socket".

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractDatabaseEngine.java [hunk 6]
```diff
@@ -252,10 +261,11 @@ protected void connect() throws Exception {
 
         final String jdbc = getFinalJdbcConnection(this.properties.getJdbc());
 
-        this.conn = DriverManager.getConnection(
-                jdbc,
-                username,
-                password);
+        final Properties props = getDBProperties();
```
Comment (lmtjalves) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractDatabaseEngine.java]: Is this change needed?

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractDatabaseEngine.java [hunk 7]
```diff
@@ -221,6 +222,14 @@ protected String getPrivateKey() throws Exception {
         return readString(new FileInputStream(f)).trim();
     }
 
+    /**
```
Comment (lmtjalves) [inline src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractDatabaseEngine.java]: WDYT of having this instead:
```
Get the properties to be passed to the database connection that will be created. Since these properties are specific to the database engine, it is up to the {@link AbstractDatabaseEngine} implementation to specify them. This is used, for instance, to set the database connection timeout.
```

Comment (TiagoMSSantos) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractDatabaseEngine.java]: Yes, that's much better. Thanks for the feedback.

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java (hunk 3 above)
Comment (TiagoMSSantos) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java]: Already fixed.

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java (hunk 4 above)
Comment (TiagoMSSantos) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java]: Already fixed.

File: src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java [hunk 8]
```diff
@@ -84,4 +84,14 @@
      * server. This is only possible on implementations that support this behavior. The default value is {@code false}.
      */
     public static final boolean DEFAULT_DISABLE_LOB_CACHING = false;
+
+    /**
+     * Default duration (in seconds) to wait for the database connection to be established.
+     */
+    public static final String DEFAULT_LOGIN_TIMEOUT = "0";
```
Comment (LuisEduardoReis) [inline src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java]: Explain in the javadoc why the default is 0 and what it means. I'm guessing it means there is no timeout for most databases.

Comment (TiagoMSSantos) [reply src/main/java/com/feedzai/commons/sql/abstraction/util/Constants.java]: Yes, you are right, 0 means no timeout. I will explain it there. Thanks for the feedback.

File: src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/h2/H2EngineSchemaTest.java [hunk 9]
```diff
@@ -128,4 +142,37 @@ public static int GetOne() {
...
+                setProperty(SCHEMA_POLICY, "create");
+                setProperty(SOCKET_TIMEOUT, Integer.toString(socketTimeout));
+                setProperty(LOGIN_TIMEOUT, Integer.toString(loginTimeout));
+            }
+        };
+
+        final DatabaseEngine de = DatabaseFactory.getConnection(properties);
+        final int connectionTimeoutInMs = de.getConnection().getNetworkTimeout();
+        // Not supported
+        assertEquals("timeout in db connection ok?", 0, connectionTimeoutInMs);
```
Comment (LuisEduardoReis) [inline src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/h2/H2EngineSchemaTest.java]: Error message could be more specific: "Is the timeout of the DB connection disabled as expected?"

File: src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/AbstractEngineSchemaTest.java [hunk 10]
```diff
@@ -420,6 +425,38 @@ public void testInsertBatchRandomValuesDoNoWorkInBinaryDoubleColumn() throws Exc
...
+                setProperty(ENGINE, config.engine);
+                setProperty(SCHEMA_POLICY, "create");
+                setProperty(SOCKET_TIMEOUT, Integer.toString(socketTimeout));
+                setProperty(LOGIN_TIMEOUT, Integer.toString(loginTimeout));
+            }
+        };
+
+        final DatabaseEngine de = DatabaseFactory.getConnection(properties);
+        final int connectionTimeoutInMs = de.getConnection().getNetworkTimeout();
+        assertEquals("timeout in db connection ok?", socketTimeout * 1000, connectionTimeoutInMs);
```
Comment (LuisEduardoReis) [inline src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/AbstractEngineSchemaTest.java]: Error message could be more specific: "Is the timeout of the DB connection the expected?"

Comment (TiagoMSSantos) [reply src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/AbstractEngineSchemaTest.java]: Yes, that's better. Thanks.

File: src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/h2/H2EngineSchemaTest.java (hunk 9 above)
Comment (TiagoMSSantos) [reply src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/h2/H2EngineSchemaTest.java]: Yes, that is a special case where the socket timeout can't be set currently for the H2 and DB2 databases. Thanks for the feedback.

File: src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java (hunk 5 above)
Comment (TiagoMSSantos) [reply src/main/java/com/feedzai/commons/sql/abstraction/engine/configuration/PdbProperties.java]: Already fixed. Thanks for the comment.
